### ISSUE TITLE
Dereferencable labels

### ISSUE THREAD

**Issue by stephenwf** [IIIF-Commons/parser#5]
Occasionally there are resources referenced in IIIF that have labels that do not match the labels on the resource after dereferencing. 

Example: [https://iiif.wellcomecollection.org/presentation/b18031511](https://iiif.wellcomecollection.org/presentation/b18031511)

In the manifest
```json
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "type": "Manifest",
  "label": {
    "en": [
      "Volume 1",
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar."
    ]
  }
}
```

When you dereference it:
```json
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "type": "Manifest",
  "label": {
    "en": [
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar."
    ]
  }
}
```

The way the vault works at the moment, the first label inside of the collection (with "Volume 1") would be saved in the vault when you import the collection. When you then did `vault.load(manifestId)` the label would be replaced with the new one (and "Volume 1" would be lost).

The issue is that there are 2 labels, one for the reference - and one for the resource.

Link to the code where references are added - https://github.com/IIIF-Commons/parser/blob/main/src/presentation-3/normalize.ts#L79-L80

**Possible fixes or improvements**

* Add labels to _some_ references like collection items or annotation pages that are known to be external
* Better merging of entities, so no values are lost (but loses contextual information of where they came from)
* Using `meta` to store supplemental labels that can be accessed if required.

**Comment by tomcrane** [comment]
The problem is one of adding different parts of the graph together and sometimes getting different facts from different parts. I think, though, that this becomes a problem for software that uses vault to solve, if we can get vault to do the following:

* All refs retain their `label` property (so they are {id, type, label}). Not just refs to external-but-still-iiif resources like AnnotationPage or Collection (or Manifest, sometimes). This makes it quicker to build UI that doesn't have to go to Vault again to get the strings to display on the parts. 
* It also means you don't have to load an external resource to find a label for it, if there is already one in the initial resource.
* It solves a problem in the current impl where if the external reference in the manifest has a label but the external resource itself does not, there's no way in Vault to see what that label in the reference is.
* If vault gets one label from the reference and another from the corresponding external resource, then that resource simply has both labels. This is consistent with the graph. The language maps are merged (if the string values differ). It might gain more keys or it might gain more strings for existing keys, last in appended.

If you are using Vault in a read-only way (e.g., in a viewer) then these rules don't really give you a problem. You can even keep track of the `label` of a resource before you did `vault.load(..)` and "pulled more information into the graph", and the `label` after, and do whatever you need to do if they are different.

The problem of splitting them out again if you want to save a part of the vault state is deferred to vault-using software. Let's see if that becomes a problem.

Scenario... a Collection Editor. It's job is to save some JSON, a IIIF Collection. Obviously does not contain embedded manifests, just refs (but obviously, useless without labels).

You add 10 manifests to the collection. You use vault to load them, validate them, grab their labels - but you're just adding refs to the Collection JSON you are creating. They bring their source labels with them.
If you edit those labels to work better as member items in a collection, you're just editing them in the collection - even if vault is modifying the manifest it loaded, you're never going to save that manifest out of **that** vault, you're only going to save the collection, to JSON.

**Comment by stephenwf** [comment]
One idea, only for representations in Vault and abstracted when serialised, could be introducing a new property on any stored resource:
```json
{
  "id": "https://example.org/manifest-1",
  "...": "...",
  "dcterms:hasPart": [
    {
      "id": "https://example.org/manifest-1",
      "type": "Canvas",
      "label": "A different label",
      "partOf": "https://example.org/collection-1"
    }
  ]
}
```
For example the above would say, when this resources is "partOf" `https://example.org/collection-1` it should use this snippet instead of the main resource. It would be completely excluded when serialising the manifest, but would appear when serialising the manifest. 

I think it would be generic enough to work across most scenarios too:
```json
{
  "id": "https://example.org/large-image.jpg/full/300,200/0/default.jpg",
  "type": "Image",
  "width": 300,
  "height": 200,
  "dcterms:hasPart": [
    {
      "id": "https://example.org/large-image.jpg/full/300,43/0/default.jpg",
      "type": "Image",
      "service": "{ ... inline image service ... }",
      "width": 300,
      "height": 200,
      "partOf": "https://example.org/painting-anno-1"
    }
  ]
}
```
This would fix a bug in Manifest Editor where if the image resource on the painting annotation is the same as the canvas thumbnail it can have different properties (maybe I don't want an image service on the thumbnail - just a single size).

**Comment by tomcrane** [comment]
> The problem of splitting them out again if you want to save a part of the vault state is deferred to vault-using software. Let's see if that becomes a problem.

Yeah I think that Vault needs to help here, contrary to my earlier comment.

However it does it, it needs to be aware of the serialisation boundary a resource came into Vault from, so that if it's serialised out again to that boundary, the correct slice of Vault gets serialised.

IIIF is a serialisation-specific spec first and graph data second.

This is like tagging _everything_ with some kind of context id of the relevant serialisation boundary (typically a Manifest id, but could be anything).

**Comment by stephenwf** [comment]
This is now implemented in: #23 

Here's roughly how it works. Using the original example where the Collection and Manifest have different values. When you first load the Collection the following will be saved:
```json
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "type": "Manifest",
  "label": {
    "en": [
      "Volume 1",
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://example.org/collection-1",
      "type": "Manifest",
    },
  ],
}
```

And if you were to load only the manifest on it's own, you would get:
```json
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "label": {
    "en": [
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "type": "Manifest",
    },
  ],
  "type": "Manifest",
}
```

But if you load both, then regardless of the order it will look like:
```json
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "type": "Manifest",
  "label": {
    "en": [
      "Volume 1",
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "type": "Manifest",
      "label": {
        "en": [
          "Volume 1",
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "iiif-parser:partOf": "https://example.org/collection-1",
    },
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "type": "Manifest",
      "label": {
        "en": [
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "iiif-parser:partOf": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
    },
  ],
}
```

During serialization this is transparent, you just get the right variation at the correct time. However for Vault this will have to change slightly to support it - with the optional `partOf` option for `.get()` and some caching considerations / documentation as the instances of the objects change each time you request them. 


### Image resources

The other example that's cropped up is image resources (inside annotations) sharing IDs with thumbnails - sometimes with different values.

This is the manifest I put together, based on a cookbook one. Note the `canvas.thumbnail` is just an image, but inside the annotation it has an image service and height and width.

```json
{
  "@context": "http://iiif.io/api/presentation/3/context.json",
  "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
  "type": "Manifest",
  "label": {
    "en": [
      "Picture of Göttingen taken during the 2019 IIIF Conference"
    ]
  },
  "items": [
    {
      "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
      "type": "Canvas",
      "label": {
        "en": [
          "Canvas with a single IIIF image"
        ]
      },
      "thumbnail": [
        {
          "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
          "type": "Image",
          "format": "image/jpeg"
        }
      ],
      "height": 3024,
      "width": 4032,
      "items": [
        {
          "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
          "type": "AnnotationPage",
          "items": [
            {
              "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
              "type": "Annotation",
              "motivation": "painting",
              "body": {
                "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
                "type": "Image",
                "format": "image/jpeg",
                "height": 3024,
                "width": 4032,
                "service": [
                  {
                    "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
                    "profile": "level1",
                    "type": "ImageService3"
                  }
                ]
              },
              "target": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1"
            }
          ]
        }
      ]
    }
  ]
}

```

The V2 branch can now correctly parse this and serialize it out back to this exact JSON (🎉 ) It does that by storing it with a JSON-LD framing:

```json
{
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "type": "Image",
  "format": "image/jpeg",
  "height": 3024,
  "width": 4032,
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "type": "Image",
      "format": {},
      "iiif-parser:partOf": "https://example.org/canvas-2",
    },
    {
      "@explicit": true,
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "type": "Image",
      "format": {},
      "height": {},
      "service": {},
      "width": {},
      "iiif-parser:partOf": "https://example.org/canvas-1",
    },
  ],
  "service": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
      "profile": "level1",
      "type": "ImageService3",
    },
  ],
}
```

So unpacking both of the frames:
```json
{
  "@explicit": true,
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "type": "Image",
  "format": {},
  "iiif-parser:partOf": "https://example.org/canvas-2",
}
```
This is what should be embedded on the canvas (thumbnail property). 

*  `@explicit` states that the resource should only include properties listed ([W3C](https://w3c.github.io/json-ld-framing/#dfn-explicit-inclusion-flag))
* `{}` is a wildcard that says "include this value" ([W3C](https://w3c.github.io/json-ld-framing/#dfn-wildcard))
* `"type": "Image"` is a literal which will just be included, this will be how differing values will be stated
* `"iiif-parser:partOf"` is where this framing applies

The other frame includes lots of wildcards. These rules are all computed when you serialize it out - and is largely non-breaking for the serializer. 

### Limitations

You could construct a case where this would break, where the `partOf` is not specific enough. A workaround could be to extend it to ` "iiif-parser:partOf": "https://example.org/canvas-2#thumbnail"` to be more specific - however I can't think of a real-world case where this might happen and it complicates the parsing.

Additionally, this will add some complexity to Manifest Editor, if it came across these types of cases. When editing, a context for where a resource is being edited will have to be added - and potentially some rules for which properties to change and when to update the "merged" state.
